File: nodejs/packages/periph/src/chips/led/apa102.js
'use strict';

/**
 * APA102 addressable RGB LED strip — minimal interface.
 *
 * Drives a chain of n APA102 pixels over an SPI connection (Mode 0, MSB first).
 * Maintains an internal BGR+brightness buffer; fill() writes every pixel and
 * transmits the full frame (start + pixels + end) immediately.
 *
 * The APA102 frame format is:
 *   start frame: 4 zero-bytes (0x00 × 4)
 *   pixel data:  n × 4 bytes [0xE0|brightness, B, G, R] (BGR wire order)
 *   end frame:   max(4, (n+15)//16) bytes of 0xFF
 *
 * @param {import('../../connection/spi').SPIConnection} connection - Configured SPI connection (Mode 0, MSB first).
 * @param {number} n         - Number of pixels in the strip.
 */
class APA102Minimal {
    /**
     * @param {import('../../connection/spi').SPIConnection} connection - Configured SPI connection (Mode 0, MSB first).
     * @param {number} n         - Number of pixels in the strip.
     */
    constructor(connection, n) {
        this._conn = connection;
        this._n = n;
        this._buf = Buffer.alloc(n * 4);
        // Initialize buffer with hardware brightness=31, all channels off
        for (let i = 0; i < n; i++) {
            this._buf[i * 4]     = 0xE0 | 31;  // brightness byte (3 high bits = 1)
            this._buf[i * 4 + 1] = 0;          // blue
            this._buf[i * 4 + 2] = 0;          // green
            this._buf[i * 4 + 3] = 0;          // red
        }
    }

    /**
     * Fill every pixel with one colour and send to the strip immediately.
     *
     * Clamps each channel to [0, 255]. Stores brightness/B/G/R in the
     * internal buffer (BGR wire order with hardware brightness byte first),
     * then transmits the full APA102 frame.
     *
     * @param {number} r - Red channel (0–255).
     * @param {number} g - Green channel (0–255).
     * @param {number} b - Blue channel (0–255).
     * @returns {Promise<void>}
     */
    async fill(r, g, b) {
        r = Math.max(0, Math.min(255, r | 0));
        g = Math.max(0, Math.min(255, g | 0));
        b = Math.max(0, Math.min(255, b | 0));
        for (let i = 0; i < this._n; i++) {
            this._buf[i * 4]     = 0xE0 | 31;  // hardware brightness = 31 (max)
            this._buf[i * 4 + 1] = b;          // blue
            this._buf[i * 4 + 2] = g;          // green
            this._buf[i * 4 + 3] = r;          // red
        }
        await this._sendFrame();
    }

    /**
     * Turn off all pixels (fill with black and send).
     *
     * Equivalent to fill(0, 0, 0).
     * @returns {Promise<void>}
     */
    async off() {
        await this.fill(0, 0, 0);
    }

    /**
     * Send the full APA102 frame (start + pixel buffer + end).
     * @returns {Promise<void>}
     * @private
     */
    async _sendFrame() {
        const endBytes = Math.max(4, Math.floor((this._n + 15) / 16));
        const frame = Buffer.alloc(4 + this._n * 4 + endBytes);
        // Start frame: 4 zero bytes
        frame[0] = frame[1] = frame[2] = frame[3] = 0x00;
        // Pixel data
        this._buf.copy(frame, 4);
        // End frame: 0xFF bytes
        for (let i = 0; i < endBytes; i++) {
            frame[4 + this._n * 4 + i] = 0xFF;
        }
        await this._conn.write(frame);
    }
}

/**
 * APA102 full interface — extends APA102Minimal with per-pixel control.
 *
 * Adds individual pixel addressing with per-pixel hardware brightness,
 * explicit show(), global software brightness scaling, buffer rotation,
 * and HSV fill. Call set_pixel() / set_pixels() to update the buffer,
 * then show() to transmit; or use the inherited fill() for an immediate
 * all-same-colour update.
 *
 * @param {import('../../connection/spi').SPIConnection} connection - Configured SPI connection (Mode 0, MSB first).
 * @param {number} n         - Number of pixels in the strip.
 */
class APA102Full extends APA102Minimal {
    /**
     * @param {import('../../connection/spi').SPIConnection} connection - Configured SPI connection (Mode 0, MSB first).
     * @param {number} n         - Number of pixels in the strip.
     */
    constructor(connection, n) {
        super(connection, n);
        this._brightness = 255;  // global software brightness (0–255)
    }

    /**
     * Global software brightness scalar applied at show() time (0–255).
     * @type {number}
     */
    get brightness() { return this._brightness; }
    set brightness(value) {
        this._brightness = Math.max(0, Math.min(255, value | 0));
    }

    /**
     * Set one pixel in the buffer without sending.
     *
     * Index is clamped to [0, n-1]; each RGB channel is clamped to [0, 255];
     * pixel_brightness is clamped to [0, 31]. Call show() to transmit.
     *
     * @param {number} index             - Zero-based pixel index.
     * @param {number} r                 - Red channel (0–255).
     * @param {number} g                 - Green channel (0–255).
     * @param {number} b                 - Blue channel (0–255).
     * @param {number} [pixel_brightness=31] - Per-pixel hardware brightness 0–31.
     */
    set_pixel(index, r, g, b, pixel_brightness = 31) {
        index = Math.max(0, Math.min(this._n - 1, index | 0));
        r = Math.max(0, Math.min(255, r | 0));
        g = Math.max(0, Math.min(255, g | 0));
        b = Math.max(0, Math.min(255, b | 0));
        pixel_brightness = Math.max(0, Math.min(31, pixel_brightness | 0));
        this._buf[index * 4]     = 0xE0 | pixel_brightness;
        this._buf[index * 4 + 1] = b;
        this._buf[index * 4 + 2] = g;
        this._buf[index * 4 + 3] = r;
    }

    /**
     * Set multiple pixels from an array of [r, g, b] or [r, g, b, pixel_brightness] arrays.
     *
     * Each element is [r, g, b] or [r, g, b, pixel_brightness]. Missing brightness
     * defaults to 31. Extra entries beyond the strip length are ignored.
     * Does not transmit — call show() afterwards.
     *
     * @param {Array<number[]>} colors - Array of [r, g, b] or [r, g, b, pixel_brightness] arrays.
     */
    set_pixels(colors) {
        for (let i = 0; i < colors.length && i < this._n; i++) {
            const [r, g, b, pixel_brightness = 31] = colors[i];
            this._buf[i * 4]     = 0xE0 | Math.max(0, Math.min(31, pixel_brightness | 0));
            this._buf[i * 4 + 1] = Math.max(0, Math.min(255, b | 0));
            this._buf[i * 4 + 2] = Math.max(0, Math.min(255, g | 0));
            this._buf[i * 4 + 3] = Math.max(0, Math.min(255, r | 0));
        }
    }

    /**
     * Transmit the current buffer to the strip, applying software brightness scaling.
     *
     * Each RGB channel value is scaled: sent = stored * brightness / 255.
     * The per-pixel hardware brightness byte is NOT scaled.
     * @returns {Promise<void>}
     */
    async show() {
        const bri = this._brightness;
        const endBytes = Math.max(4, Math.floor((this._n + 15) / 16));
        const pixelDataLen = this._n * 4;
        const totalLen = 4 + pixelDataLen + endBytes;

        const frame = Buffer.alloc(totalLen);
        frame[0] = frame[1] = frame[2] = frame[3] = 0x00;

        if (bri === 255) {
            this._buf.copy(frame, 4);
        } else {
            // Scale RGB channels, leave hardware brightness byte unchanged
            for (let i = 0; i < this._n; i++) {
                const base = i * 4;
                frame[4 + base]     = this._buf[base];                                         // hardware brightness
                frame[4 + base + 1] = (this._buf[base + 1] * bri / 255) | 0;  // blue
                frame[4 + base + 2] = (this._buf[base + 2] * bri / 255) | 0;  // green
                frame[4 + base + 3] = (this._buf[base + 3] * bri / 255) | 0;  // red
            }
        }

        for (let i = 0; i < endBytes; i++) {
            frame[4 + pixelDataLen + i] = 0xFF;
        }

        await this._conn.write(frame);
    }

    /**
     * Shift the pixel buffer left by steps whole-pixel positions (wraps around).
     *
     * Each step shifts 4 bytes (one BGR+brightness pixel). Does not transmit —
     * call show() afterwards.
     *
     * @param {number} [steps=1] - Number of pixel positions to shift left.
     */
    rotate(steps = 1) {
        steps = ((steps % this._n) + this._n) % this._n;
        if (steps === 0) return;
        const bytes = steps * 4;
        const head = Buffer.from(this._buf.slice(0, bytes));
        this._buf.copyWithin(0, bytes);
        head.copy(this._buf, this._buf.length - bytes);
    }

    /**
     * Fill every pixel with one HSV colour and send to the strip immediately.
     *
     * Converts HSV to RGB then calls fill() at hardware brightness 31.
     *
     * @param {number} h - Hue (0.0–1.0).
     * @param {number} s - Saturation (0.0–1.0).
     * @param {number} v - Value/brightness (0.0–1.0).
     * @returns {Promise<void>}
     */
    async fill_hsv(h, s, v) {
        const [r, g, b] = _hsvToRgb(h, s, v);
        await this.fill(r, g, b);
    }
}

const { _hsvToRgb } = require('./_color');

module.exports = { APA102Minimal, APA102Full };

File: python/periph/chips/led/apa102.py
from ._color import _hsv_to_rgb


class APA102Minimal:
    """APA102 addressable RGB LED strip — minimal interface.

    Drives a chain of n APA102 pixels over an SPI connection.
    Maintains an internal BGR+brightness buffer; fill() writes all pixels
    and transmits immediately with start/end framing.

    Args:
        connection: Configured SPI connection (MicroPython, CircuitPython, or Linux).
        n: Number of pixels in the strip.
    """

    def __init__(self, connection, n):
        """Initialise APA102Minimal with a connection and pixel count.

        Args:
            connection: Configured SPI connection.
            n: Number of pixels in the strip (must be >= 1).
        """
        self._connection = connection
        self._n = n
        self._buf = bytearray(n * 4)
        for i in range(n):
            self._buf[i * 4] = 0xE0 | 31  # brightness = 31 (max)
            self._buf[i * 4 + 1] = 0      # blue
            self._buf[i * 4 + 2] = 0      # green
            self._buf[i * 4 + 3] = 0      # red

    def fill(self, r, g, b):
        """Fill every pixel with one colour and send to the strip immediately.

        Clamps each channel to [0, 255]. Stores brightness/B/G/R in the
        internal buffer (BGR wire order with hardware brightness byte first)
        then transmits the full frame (start + pixels + end).

        Args:
            r: Red channel (0–255).
            g: Green channel (0–255).
            b: Blue channel (0–255).
        """
        r = max(0, min(255, int(r)))
        g = max(0, min(255, int(g)))
        b = max(0, min(255, int(b)))
        for i in range(self._n):
            self._buf[i * 4] = 0xE0 | 31  # hardware brightness = 31
            self._buf[i * 4 + 1] = b       # blue
            self._buf[i * 4 + 2] = g       # green
            self._buf[i * 4 + 3] = r       # red
        self._write_frame()

    def off(self):
        """Turn off all pixels (fill with black and send).

        Equivalent to fill(0, 0, 0).
        """
        self.fill(0, 0, 0)

    def _write_frame(self):
        """Send the full APA102 frame (start + pixel buffer + end)."""
        start_frame = bytes(4)  # 0x00 × 4
        end_bytes = max(4, (self._n + 15) // 16)
        end_frame = bytes([0xFF] * end_bytes)
        self._connection.write(start_frame + bytes(self._buf) + end_frame)


class APA102Full(APA102Minimal):
    """APA102 full interface — extends APA102Minimal with per-pixel control.

    Adds individual pixel addressing with per-pixel hardware brightness,
    explicit show(), global software brightness scaling, buffer rotation,
    and HSV fill. Call set_pixel() / set_pixels() to update the buffer,
    then show() to transmit; or use the inherited fill() for an immediate
    all-same-colour update.

    Args:
        connection: Configured SPI connection.
        n: Number of pixels in the strip.
    """

    def __init__(self, connection, n):
        """Initialise APA102Full with a connection and pixel count.

        Args:
            connection: Configured SPI connection.
            n: Number of pixels in the strip.
        """
        super().__init__(connection, n)
        self._brightness = 255  # global software brightness (0–255)

    @property
    def brightness(self):
        """Global brightness scalar applied at show() time (0–255)."""
        return self._brightness

    @brightness.setter
    def brightness(self, value):
        self._brightness = max(0, min(255, int(value)))

    def set_pixel(self, index, r, g, b, pixel_brightness=31):
        """Set one pixel in the buffer without sending.

        Clamps index to [0, n-1], RGB channels to [0, 255], and
        hardware brightness to [0, 31]. Call show() to transmit.

        Args:
            index: Zero-based pixel index.
            r: Red channel (0–255).
            g: Green channel (0–255).
            b: Blue channel (0–255).
            pixel_brightness: Per-pixel hardware brightness 0–31 (default 31).
        """
        index = max(0, min(self._n - 1, int(index)))
        r = max(0, min(255, int(r)))
        g = max(0, min(255, int(g)))
        b = max(0, min(255, int(b)))
        pixel_brightness = max(0, min(31, int(pixel_brightness)))
        self._buf[index * 4] = 0xE0 | pixel_brightness
        self._buf[index * 4 + 1] = b
        self._buf[index * 4 + 2] = g
        self._buf[index * 4 + 3] = r

    def set_pixels(self, colors):
        """Write a sequence of color tuples into the buffer starting at pixel 0.

        Each element is (r, g, b) or (r, g, b, pixel_brightness).
        Missing brightness defaults to 31. Extra entries beyond the strip
        length are ignored. Call show() to transmit.

        Args:
            colors: Iterable of (r, g, b) or (r, g, b, pixel_brightness) tuples.
        """
        for i, color in enumerate(colors):
            if i >= self._n:
                break
            r, g, b = color[0], color[1], color[2]
            pixel_brightness = color[3] if len(color) > 3 else 31
            r = max(0, min(255, int(r)))
            g = max(0, min(255, int(g)))
            b = max(0, min(255, int(b)))
            pixel_brightness = max(0, min(31, int(pixel_brightness)))
            self._buf[i * 4] = 0xE0 | pixel_brightness
            self._buf[i * 4 + 1] = b
            self._buf[i * 4 + 2] = g
            self._buf[i * 4 + 3] = r

    def show(self):
        """Transmit the current buffer to the strip, applying software brightness scaling.

        Each RGB channel value is scaled: sent = stored * brightness // 255.
        The per-pixel hardware brightness byte is not affected by this scaling.
        """
        bri = self._brightness
        start_frame = bytes(4)
        end_bytes = max(4, (self._n + 15) // 16)
        end_frame = bytes([0xFF] * end_bytes)

        if bri == 255:
            self._connection.write(start_frame + bytes(self._buf) + end_frame)
        else:
            scaled = bytearray(len(self._buf))
            for i in range(self._n):
                base = i * 4
                # Hardware brightness byte unchanged
                scaled[base] = self._buf[base]
                # Scale RGB channels
                scaled[base + 1] = self._buf[base + 1] * bri // 255  # blue
                scaled[base + 2] = self._buf[base + 2] * bri // 255  # green
                scaled[base + 3] = self._buf[base + 3] * bri // 255  # red
            self._connection.write(start_frame + bytes(scaled) + end_frame)

    def rotate(self, steps=1):
        """Shift the pixel buffer left by steps positions (wraps around).

        Operates on whole 4-byte pixel units. Does not transmit — call show()
        afterwards.

        Args:
            steps: Number of pixel positions to shift left (default 1).
        """
        steps = steps % self._n
        if steps == 0:
            return
        s4 = steps * 4
        self._buf = self._buf[s4:] + self._buf[:s4]

    def fill_hsv(self, h, s, v):
        """Fill every pixel with one HSV colour and send to the strip immediately.

        Converts HSV to RGB then calls fill() at hardware brightness 31.

        Args:
            h: Hue (0.0–1.0).
            s: Saturation (0.0–1.0).
            v: Value/brightness (0.0–1.0).
        """
        r, g, b = _hsv_to_rgb(h, s, v)
        self.fill(r, g, b)